src/middleware/cookies.rs

```diff
@@ -0,0 +1,274 @@
+    pub fn from_headers(headers: &HeaderMap) -> Self {
+        CookieData {
+            content: Arc::new(RwLock::new(
+                headers
+                    .get(http::header::COOKIE)
+                    .and_then(|raw| parse_from_header(raw.to_str().unwrap()).ok())
+                    .unwrap_or_default(),
+            )),
+        }
+    }
+}
+
+fn parse_from_header(s: &str) -> Result<CookieJar, ParseError> {
+    let mut jar = CookieJar::new();
+
+    s.split(';').try_for_each(|s| -> Result<_, ParseError> {
+        jar.add_original(Cookie::parse(s.trim().to_owned())?);
+        Ok(())
+    })?;
+
+    Ok(jar)
+}
+
+/// An extension to `Context` that provides cached access to cookies
+pub trait RequestExt {
+    /// returns a `Cookie` by name of the cookie
+    fn get_cookie(&self, name: &str) -> Result<Option<Cookie<'static>>, Error>;
+
+    /// Add cookie to the cookie jar
+    fn set_cookie(&mut self, cookie: Cookie<'static>) -> Result<(), Error>;
+
+    /// Removes the cookie. This instructs the `CookiesMiddleware` to send a cookie with empty value
+    /// in the response.
+    fn remove_cookie(&mut self, cookie: Cookie<'static>) -> Result<(), Error>;
+}
+
+impl<State> RequestExt for Request<State> {
+    fn get_cookie(&self, name: &str) -> Result<Option<Cookie<'static>>, Error> {
+        let cookie_data = self
+            .local::<CookieData>()
+            .ok_or_else(|| StringError(MIDDLEWARE_MISSING_MSG.to_owned()))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+        let locked_jar = cookie_data
+            .content
+            .read()
+            .map_err(|e| StringError(format!("Failed to get read lock: {}", e)))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+        Ok(locked_jar.get(name).cloned())
+    }
+
+    fn set_cookie(&mut self, cookie: Cookie<'static>) -> Result<(), Error> {
+        let cookie_data = self
+            .local::<CookieData>()
+            .ok_or_else(|| StringError(MIDDLEWARE_MISSING_MSG.to_owned()))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+        let mut locked_jar = cookie_data
+            .content
+            .write()
+            .map_err(|e| StringError(format!("Failed to get read lock: {}", e)))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+        locked_jar.add(cookie);
+        Ok(())
+    }
+
+    fn remove_cookie(&mut self, cookie: Cookie<'static>) -> Result<(), Error> {
+        let cookie_data = self
+            .local::<CookieData>()
+            .ok_or_else(|| StringError(MIDDLEWARE_MISSING_MSG.to_owned()))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+        let mut locked_jar = cookie_data
+            .content
+            .write()
+            .map_err(|e| StringError(format!("Failed to get read lock: {}", e)))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+        locked_jar.remove(cookie);
+        Ok(())
+    }
+}
```

Comment:
These should probably be inherent methods on `Request`. We've moved away from providing `Ext` traits on `Request` / `Response` in Tide since we've defined our own.